setup.py

```diff
@@ -20,41 +104,119 @@ def get_extensions():
+                     extra_link_args=[],
+                     language='c++')
+                     #include_dirs = [get_include])
+
+    ext3 = Extension(name='commit.proximals',
+                      sources=['commit/proximals.pyx'],
+                      extra_compile_args= {'gcc':  ['-w'],
+                                           'nvcc': ['-arch=sm_50', '--ptxas-options=-v', '-c', '--compiler-options', "'-fPIC'"]},
+                      extra_link_args=[],
+                      language='c++')
+                      #include_dirs = [get_include])
+
+    ext4 = Extension(name='commit.cudaoperator',
+                     sources = ['commit/operator_withCUDA.cu', 'commit/cudaoperator.pyx'],
+                     extra_compile_args= {'gcc':  ['-w'],
+                                          'nvcc': ['-arch=sm_50', '--ptxas-options=-v', '-c', '--compiler-options', "'-fPIC'"]},
+                     language = 'c++',
+                     #include_dirs = [get_include, CUDA['include']],
+                     library_dirs = [CUDA['lib64']],
+                     libraries = ['cudart'],
+                     runtime_library_dirs = [CUDA['lib64']])
+
+    return [ext1, ext2, ext3, ext4]
 
-class CustomBuildExtCommand(build_ext):
-    """ build_ext command to use when numpy headers are needed. """
+if CUDA == None:
+    extensions = get_extensions()
+else:
+    extensions = get_extensions_with_cuda()
 
-    def run(self):
-        # Now that the requirements are installed, get everything from numpy
-        from Cython.Build import cythonize
-        from numpy import get_include
-        
-        # Add everything requires for build
-        self.swig_opts = None
-        self.include_dirs = [get_include()]
-        self.distribution.ext_modules[:] = cythonize(
-                    self.distribution.ext_modules)
+if CUDA == None:
+    class CustomBuildExtCommand(build_ext):
+        """ build_ext command to use when numpy headers are needed. """
 
-        # Call original build_ext command
-        build_ext.finalize_options(self)
-        build_ext.run(self)
+        def run(self):
+            # Now that the requirements are installed, get everything from numpy
+            from Cython.Build import cythonize
+            from numpy import get_include
+            
+            # Add everything requires for build
+            self.swig_opts = None
+            self.include_dirs = [get_include()]
+            self.distribution.ext_modules[:] = cythonize(self.distribution.ext_modules)
 
+            # Call original build_ext command
+            build_ext.finalize_options(self)
+            build_ext.run(self)
+else:
+    class CustomBuildExtCommand(build_ext):
+        """ build_ext command to use when numpy headers are needed. """
+
+        def build_extensions(self):
+            customize_compiler_for_nvcc(self.compiler)
+            build_ext.build_extensions(self)
```

Comment:
Is this function used?

Reply:
Yes, it is used to create a custom builder for the .cu files in Cython. It is only added when compiling the GPU version.

Reply:
OK